absl/absl_test.cpp

```diff
@@ -24,36 +24,153 @@
 namespace nda {
 
 TEST(absl_stringify_interval) {
-  interval<> i0;
   {
+    interval<> i0;
     const std::string s = absl::StrFormat("%v", i0);
     ASSERT(s == "[0, 1)");
   }
 
-  interval<> i1(-6, 11);
   {
+    interval<> i1(-6, 11);
     const std::string s = absl::StrFormat("%v", i1);
     ASSERT(s == "[-6, 5)");
   }
 
-  interval<3> i2;
   {
+    interval<3> i2;
     const std::string s = absl::StrFormat("%v", i2);
     ASSERT(s == "[3, 4)");
   }
 
-  interval<4> i3;
-  i3.set_extent(6);
   {
+    interval<4> i3;
+    i3.set_extent(6);
     const std::string s = absl::StrFormat("%v", i3);
     ASSERT(s == "[4, 10)");
   }
 
-  interval<-5, -12> i4;
   {
+    interval<-5, -12> i4;
     const std::string s = absl::StrFormat("%v", i4);
     ASSERT(s == "[-5, -17)");
   }
 }
 
+TEST(absl_stringify_dim) {
+  // Fully dynamic but unset.
+  {
+    dim<> d;
+    const std::string s = absl::StrFormat("%v", d);
+    ASSERT(s == "dim(0, 0, -9)");
```

Comment:
Maybe stringify for dims should catch `dynamic` and write that instead of `-9`?

Reply:
Ha, you fell for the same trap I did. It's not stringifying the compile-time stride `dynamic`, but the runtime value `-9`.

Reply:
I think that 'dynamic' is the right string to show here. This will allow us to decode shapes that are created as input to the array constructor as "dim(0, 256, dynamic)" for example. This is clearer than printing -9, I think.

Reply:
I think Kevin is right that this really is the value `-9`. The thing that is special about that is not that it's dynamic, but that it is almost guaranteed to be replaced by an inferred value at some point, but we don't have a token for that. There *was* `UNK` which makes more sense in this context.

Reply:
We might be able to find a better name. I just think -9 is even more obscure. The string 'unknown' instead of UNK? 'dynamic' doesn't seem too bad if we do need to choose a string in my opinion as that's the name of the constant one would use to compare against (or call `is_dynamic(dim0.stride())` on, etc.)

Reply:
@fbleibel-g Is the situation clear now? I'm just printing its runtime value, which is -9 when default initialized. I don't think it makes sense to check if that dim is dynamic and print `UNK` or `dynamic` when I'm just printing the values, not the type.

Reply:
There are two shape types - resolved and unresolved. For strides only, the runtime value nda::dynamic has a special meaning for unresolved dims when creating arrays or array_ref(base_ptr, shape), and we check it in:

```
NDARRAY_HOST_DEVICE void resolve_unknown_strides(AllDims& all_dims, Dim0& dim0, Dims&... dims) {
  if (is_dynamic(dim0.stride())) {
 ```
 
It's possible to create a shape with a runtime stride value of -9, but as soon as you make an array or array_ref out of it, it gets resolved and the value ignored. This is why I thought it might be useful to replace it with a string like "ignored", "unknown", "unspecified",... as our magic value is more of an implementation detail anyway.

Reply:
I think for now this MR is fine as-is. I want to make a change here, which is to make it so the default value for runtime strides is not the compile-time `Stride` (which is `dynamic` = -9 by default), but is instead a different magic constant `unknown` (but maybe still has the value -9, but it doesn't have to be) and then it would make more sense to catch it here when stringifying. I would have preferred a more unreasonable value (e.g. -2^31) but that made compiler errors ridiculous. But that issue only applies to compile-time constants.

Reply:
(Note that the **runtime** value itself checked by resolve_shape - this is the only way we can give nda:: our special marker)

Reply:
This all makes sense. Whether a shape has resolved or unresolved strides is a context-dependent detail that doesn't have as much to do with stringifying right now. I do like Dillon's proposal though. We can make that change and update the stringify tests accordingly.

Reply:
Dillon's change won't substantially change the problem - we'll still have a magic constant to compare against. We could use nda::dynamic for now until we make that change. I do think is more useful for a human to see anything other than -9 (our implementation detail - which will fail the tests when we change it) through absl::Stringify, whether you intend to use the Dim to create an array or not. Anyway, fine as-is.